### ISSUE TITLE
Migrate to go module

### ISSUE THREAD

**Issue by honnix** [lyft/flyte#129]
##  Why

* everything of flyte is on go-1.13 and go module has become mature to use in production
* it seems to be the future
* one fewer tool (dep) to install and manage
* no more `vendor` folder
* I tried the migration and it was smooth https://github.com/lyft/datacatalog/pull/21

**Comment by wild-endeavor** [comment]
We should definitely do this, it's just a question of when and how to get everyone on our team on the same page.  It's pretty high priority I think, if only for conformity to major K8s/go projects.  I'll try to look into it soon.

**Comment by honnix** [comment]
Awesome! Thanks.

**Comment by honnix** [comment]
To collect related PRs:

* lyft/datacatalog#21
* lyft/flytepropeller#38
* lyft/flyteadmin#35
* lyft/flyteplugins#36

**Comment by honnix** [comment]
I think I have make everything work (or will be working). Shall we get it going before new dependency updates introduced (e.g. https://github.com/lyft/flyteadmin/pull/37)?

**Comment by wild-endeavor** [comment]
Yes, we should.  I want to look at the whole things one more time, across all the repos.  Are there still issues with flytepropeller and flytestdlib?  Or have those been resolved?  I think the pflags change is still outstanding on our side.

**Comment by honnix** [comment]
Absolutely. All the problems have been resolved. Flyteidl needs a new build of pflags to work.

**Comment by honnix** [comment]
go module bugs and limitations are a bit more than i anticipated. :(

**Comment by wild-endeavor** [comment]
Discussed some more with @EngHabu on Friday.  It seems that the primary goal of both your alternate [change](https://github.com/lyft/flyteidl/pull/33/files), and the change going into go 1.14 is the ability to install specific versions of go tools (like lint, enumer, mockery, etc.).  If this is the case, can we not just avoid all this hassle by just pinning the version of each of the tools when running `go get` (and running it in an empty directory)?

**Comment by honnix** [comment]
Sure we can do that. The benefit having everything in one file is easier to track version changes and less error prone (think about `go get foo@v1` in one script and `go get foo@v2` in another). But yes this is a trade-off.

**Comment by wild-endeavor** [comment]
@jonathanburns would you be okay with that going into boilerplate?

**Comment by honnix** [comment]
_During the work migrating from dep to go module, there are a few substantial issues that need to be documented._

_There is a nice [summary](https://github.com/golang/go/issues/34025) on GitHub worth reading as well because most of the issues I have also personally experienced._

# Missing global flag to install tools for CI alike purpose

Very often support tools such as golangci-lint and mockery are needed to compile and apply checks, however due to [cmd/go: offer a consistent "global install" command · Issue #30515 · golang/go](https://github.com/golang/go/issues/30515) and there is no good way to install such tools without modifying `go.mod` and `go.sum`.

In comparison, other languages such as Javascript has `npm install -g`.

A possible workaround is illustrated by https://github.com/lyft/flytestdlib/pull/51. However things get worse when the need of forking comes, for example: https://github.com/lyft/flyteidl/pull/27/files#r359164560

Another way going forward is to add support tools in `go.mod` and `go.sum` as well. However this suffers from `go mod tidy` because it removes those since they are not referenced by any go code. Requiring contributors not to use `go mod tidy` doesn’t seem to be an option, and carefully reviewing `go.mod` and `go.sum` to make sure they are not deleted accidentally kinda defeats the purpose using go module.

There was an attempted fix https://golang.org/cl/203279 but was abandoned in favour of https://go-review.googlesource.com/c/tools/+/211538/ which addressed a variety of issues covered by [cmd/go: flags to control changes to go.mod, go.sum · Issue #34506 · golang/go](https://github.com/golang/go/issues/34506). With this fix, one could maintain two sets of `go.mod` and `go.sum` and use `go get -modfile build/go.mod -sumfile build/go.sum golangci-lint` for example. The fix is scheduled to be released as part of `go 1.14` that has its first beta out https://github.com/golang/go/releases/tag/go1.14beta1 already.

This is a pretty good fix in my opinion! We could maintain two sets of `go.mod` and `go.sum`, one for real code module management, one for support tools. I think if we don't want to wait for `go 1.14`, we can start maintaining those two sets and cd to the dir manually to install. However that means the set for support tools needs to be hand crafted and one can not run `go mod tidy`. As an example https://github.com/lyft/flyteidl/pull/33. A possible hack was described in https://github.com/golang/go/issues/25922 to make sure `go mod tidy` not wipe out everything. Shown by https://github.com/lyft/flyteidl/pull/33/commits/8ba0a6ae3277c5c9885cd1de02e32d4bb426d6b1

# Inconsistency changes of go.sum

This boils down to [cmd/go: flags to control changes to go.mod, go.sum · Issue #34506 · golang/go](https://github.com/golang/go/issues/34506), although it is still weird `go mod tidy` does not clean up accordingly.

As illustrated by https://github.com/lyft/flyteidl/pull/27/commits/bdd8117a9b263ed87a350922ab7d734831bdb60b, `go.sum` file is changed in an inconsistent way: TravisCI wanted those file tree checksum (such as `github.com/Azure/azure-sdk-for-go v10.2.1-beta+incompatible h1:/x4W7ZQV4PHJYnLUgKubojM8T+zlFEDdaBazAnA/QCY=`) but not my computer TM. On my computer `go mod tidy` is fine with either having it or not.  ¯\_(ツ)_/¯ It turns out there is an additional which `pflags || (go get github.com/lyft/flytestdlib/cli/pflags)` executed in CI and that results change of `go.sum`.

The purpose of having two checksums is explained by [Module authentication using go.sum](https://tip.golang.org/cmd/go/#hdr-Module_authentication_using_go_sum).

# Some part of go doesn’t not work with go module

As illustrated by [this change](https://github.com/lyft/flytestdlib/pull/50/commits/1d04a9db0052f6d6bcca83f46a94ba72a3b823ea), `importer` doesn’t work with go module and the official recommendation is to use `golang.org/x/tools/go/packages` instead.

There could be other things not working in a similar way, but hopefully we will not need to touch those.

**Comment by wild-endeavor** [comment]
@honnix do you think you could change the idl pr one last time to just separate out the go gets into a separate script?  Or alternatively make the change from this branch https://github.com/lyft/boilerplate/pull/4 and resubmit.

**Comment by honnix** [comment]
@wild-endeavor I have a few open comments on lyft/boilerplate#4 . Should we get those addressed first?

**Comment by EngHabu** [comment]
Awesome! I'm so glad we reached some conclusion here and that it's in boilerplate! are you in the process of updating all repos to use that?

**Comment by wild-endeavor** [comment]
I don't know if we have a solution to the pinning & replacement.  At least in the brief amount of time I looked at it, I could not get `go get` to install the version of mockery in your fork @EngHabu

**Comment by honnix** [comment]
I tried that before but no success. @wild-endeavor

**Comment by honnix** [comment]
How is it going with this?

**Comment by honnix** [comment]
I did another round of rebasing and cleanup, as well as incorporating boilerplate changes. I think we should really get this going. It's painful to catch up master changes. :(

**Comment by wild-endeavor** [comment]
Yes, I know it's super painful to keep updating everything all the time.  I'm very sorry.  We've been super behind with our own review cycle.  My only concern is that the boilerplate changes in the PR won't pick this up right? https://github.com/lyft/flyteidl/pull/27/files#diff-37aff102a57d3d7b797f152915a6dc16R31

@EngHabu and I think we should actually opt for https://github.com/lyft/flyteidl/pull/33/files.  This is the most stable/flexible approach.  As you mentioned, it's the same approach that will be part of 1.14, allows the replace statement construct, pin versions, and allows us to run `go mod tidy` if we want to without removing indirect tooling requirements.

If we're in agreement, I can help propagate the changes to boilerplate.

CC also @jonathanburns

**Comment by honnix** [comment]
I would vote for https://github.com/lyft/flyteidl/pull/33 as well.

**Comment by wild-endeavor** [comment]
Just to mention here, https://github.com/lyft/boilerplate/pull/4 was merged earlier today.  @honnix could you help me pull these changes in?  I've already pushed a couple commits to your forked PR of [flyteadmin](https://github.com/honnix/flyteadmin/commits/go-module-migration) and [flyteidl](https://github.com/honnix/flyteidl/tree/go-module-migration) but would you mind doing the others?

(i messed up the flyteidl one a bit which is why there are four commits)

**Comment by honnix** [comment]
All PRs updated according to latest boilerplate and green now.

**Comment by honnix** [comment]
As we are approaching the end of this journey, I think it makes sense to summarise the changes.

tl;dr, this is all about simulating `go get -modfile build/go.mod -sumfile build/go.sum golangci-lint` which will be included in `go-1.14`.

### Details

* There is a separated set of `go.mod` and `go.sum` offered by [golang_support_tools](https://github.com/lyft/boilerplate/tree/master/boilerplate/lyft/golang_support_tools)
* [tools.go](https://github.com/lyft/boilerplate/blob/master/boilerplate/lyft/golang_support_tools/tools.go) is used to force go module keeping track of the packages, so that `go mod tidy` wouldn't wipe out everything
* [download_tooling.sh](https://github.com/lyft/boilerplate/blob/master/boilerplate/lyft/golang_test_targets/download_tooling.sh) is used by a `Makefile` target (`make download_tooling`) to download all the tools; to be extra careful, we first copy `go.mod` and `go.sum` to a temp folder and then install the tools from there
* Every time we bump of version of a tool, we will run `go mod tidy` to get new checksum in place

**Comment by wild-endeavor** [comment]
As of today, we have migrated all golang Flyte repos to use go mod. Thanks to @honnix for driving this change.

Boilerplate changes: https://github.com/lyft/boilerplate/pull/4
The boilerplate code will now default to go.mod.

https://github.com/lyft/flytestdlib/releases/tag/v0.3.0 (was already on go mod as of v0.2.29)
https://github.com/lyft/flyteidl/releases/tag/v0.17.0
https://github.com/lyft/flyteplugins/releases/tag/v0.3.0
https://github.com/lyft/flytepropeller/releases/tag/v0.2.0
https://github.com/lyft/flyteadmin/releases/tag/v0.2.0
https://github.com/lyft/datacatalog/releases/tag/0.2.0

We’ve introduced a minor workaround to the boilerplate repo (and hence to all other go repos).  Please read through the thorough background discussion that @honnix posted around go modules in the GitHub issue (https://github.com/lyft/flyte/issues/129). In order to persist and standardize on a version of the common golang tools, we’ve added a second set of go.mod/sum files. These are listed in the boilerplate PR and basically separate your code from these tools. That is, your code likely uses golangci-lint but that doesn’t mean it should be part of your project’s go.mod files. Honnix’s comment has more information. This dual go.mod approach is also slated to be introduced more formally and rigorously in go 1.14.

This is a good overview of the new commands:
https://github.com/golang/go/wiki/Modules#daily-workflow

Note also that in most repos, we’re now pinning the version of client-go to something compatible with the Lyft fork of K8s (see https://github.com/lyft/datacatalog/pull/21#issuecomment-579411660 for more information).
